CODING_GUIDELINES.md

```diff
@@ -0,0 +1,64 @@
+# Coding Guidelines
+
+This document is an extension to [CONTRIBUTING](./CONTRIBUTING.md) and provides more details about the coding guidelines and requirements.
+
+## API & Design
+
+- Code must be well structured:
+  - packages must have a limited responsibility (different concerns can go to different packages),
+  - types must be easy to compose,
+  - think about maintainbility and testability.
+- "Depend upon abstractions, [not] concretions".
+- Try to limit the number of methods you are exposing. It's easier to expose something later than to hide it.
+- Take advantage of `internal` package concept.
+- Follow agreed-upon design patterns and naming conventions.
```

Comment:
@jelysn I think we need to clarify that maybe link to the golang coding best practices or borrow something from the cosmos dev community.

Reply:
Added reference link